File: CIMWeb/src/test/java/ch/ninecode/cim/cimweb/CIMWebTest.java
package ch.ninecode.cim.cimweb;

import static org.junit.Assert.assertNotNull;

import java.io.File;
import java.util.Properties;

import javax.inject.Inject;
import javax.naming.Context;
import javax.naming.InitialContext;
import javax.naming.NameClassPair;
import javax.naming.NamingEnumeration;
import javax.naming.NamingException;
import javax.resource.cci.Connection;
import javax.resource.cci.ConnectionFactory;

import org.jboss.arquillian.container.test.api.Deployment;
import org.jboss.arquillian.junit.Arquillian;
import org.jboss.shrinkwrap.api.ShrinkWrap;
import org.jboss.shrinkwrap.api.asset.FileAsset;
import org.jboss.shrinkwrap.api.exporter.ZipExporter;
import org.jboss.shrinkwrap.api.spec.EnterpriseArchive;
import org.jboss.shrinkwrap.api.spec.WebArchive;
import org.junit.Test;
import org.junit.runner.RunWith;

/**
 * CIMWeb test suite.
 */
@RunWith (Arquillian.class)
public class CIMWebTest
{
    private static final String WEBEAR_SRC = "src/main";
    private static final String WEBAPP_SRC = WEBEAR_SRC + "/webapp";

    @Deployment
    public static EnterpriseArchive createDeployment ()
    {
        final WebArchive war = ShrinkWrap.create (WebArchive.class, "CIMWeb.war");
        war.setWebXML (new File (WEBAPP_SRC, "WEB-INF/web.xml"));
        war.addPackage (java.lang.Package.getPackage ("ch.ninecode.cim.cimweb")); // getClass().getPackage()
        war.deleteClass (CIMWebTest.class);
        war.addAsWebResource (new File (WEBAPP_SRC, "index.html"));
        war.addManifest ();
        System.out.println (war.toString (true));
        war.as (ZipExporter.class).exportTo (new File ("./target/CIMWeb.war"), true);

        final EnterpriseArchive ear = ShrinkWrap.create (EnterpriseArchive.class, "CIMWeb.ear");
        ear.addAsModules (war);
        ear.add (new FileAsset (new File ("../CIMConnector/target/CIMConnector-1.0-SNAPSHOT.rar")), "CIMConnector.rar");
        ear.addManifest ();
        ear.addAsManifestResource (new File (WEBEAR_SRC, "application.xml"));

        System.out.println (ear.toString (true));
        ear.as (ZipExporter.class).exportTo (new File ("./target/CIMWeb.ear"), true);

        return (ear);
    }

    Context context;

    void print_context_r (String name, int depth)
    {
        try
        {
            NamingEnumeration<NameClassPair> values = context.list (name);
            while (values.hasMore ())
            {
                NameClassPair pair = values.next ();
                for (int i = 0; i < depth; i++)
                    System.out.print ("    ");
                System.out.println (pair.getName () + " : " + pair.getClassName ());
                if ("org.apache.openejb.core.ivm.naming.IvmContext" == pair.getClassName ())
                    print_context_r (name + "/" + pair.getName (), depth + 1);
            }
        }
        catch (NamingException e)
        {

        }
    }

    void print_context (String name)
    {
        System.out.println (name);
        print_context_r (name, 1);
    }

    @Inject
    SimpleRESTEJB ejbobject;

    @Test
    public void testResourceAdapter () throws Exception
    {
        final Properties properties = new Properties ();
        properties.setProperty (Context.INITIAL_CONTEXT_FACTORY, "org.apache.openejb.client.LocalInitialContextFactory");
        properties.setProperty ("openejb.deployments.classpath.include", ".*resource-injection.*");
        final InitialContext initialContext = new InitialContext (properties);
        context = initialContext;
        print_context ("java:");
        print_context ("openejb:");
        final ConnectionFactory connectionFactory = (ConnectionFactory) initialContext.lookup ("openejb:Resource/CIMConnector.rar");
        final Connection connection = connectionFactory.getConnection ();
        assertNotNull (connection);
        connection.close ();
    }
//    assertTrue ("interaction returned false", interaction.execute (spec, input, output));
//    assertFalse ("interaction returned empty", output.isEmpty ());
//    assertEquals ("interaction returned wrong value", "Hello World!", output.get (0).toString ());

    @Test
    public void testWebApp () throws Exception
    {
        assertNotNull (ejbobject);
        String text = ejbobject.ejb ();
        assertNotNull (text);
        System.out.println (text);
    }
}

File: CIMWeb/src/test/scala/ch/ninecode/cim/cimweb/CIMWebTest.scala
package ch.ninecode.cim.cimweb

import java.io.File
import java.util.Properties

import javax.naming.Context
import javax.naming.InitialContext
import javax.naming.NamingException
import javax.resource.cci.ConnectionFactory

import org.jboss.arquillian.container.test.api.Deployment
import org.jboss.arquillian.junit.Arquillian
import org.jboss.shrinkwrap.api.ShrinkWrap
import org.jboss.shrinkwrap.api.asset.FileAsset
import org.jboss.shrinkwrap.api.exporter.ZipExporter
import org.jboss.shrinkwrap.api.spec.EnterpriseArchive
import org.jboss.shrinkwrap.api.spec.WebArchive

import org.junit.Assert.assertNotNull
import org.junit.Assert.fail
import org.junit.Test
import org.junit.runner.RunWith

/**
 * CIMWeb test suite.
 */
@RunWith (classOf[Arquillian])
object CIMWebTest
{
    val WEBEAR_SRC = "src/main"
    val WEBAPP_SRC = s"$WEBEAR_SRC/webapp"
    println (WEBAPP_SRC)

    @Deployment
    def createDeployment: Option[EnterpriseArchive] =
    {
        try
        {
            val war = ShrinkWrap.create (classOf[WebArchive], "CIMWeb.war")
                .setWebXML (new File (WEBAPP_SRC, "WEB-INF/web.xml"))
                .addPackage (getClass.getPackage) // deprecated: (java.lang.Package.getPackage ("ch.ninecode.cim.cimweb"))
                .deleteClass (classOf[CIMWebTest])
                .addAsWebResource (new File (WEBAPP_SRC, "index.html"))
                .addManifest ()
            println (war.toString (true))
            war.as (classOf[ZipExporter]).exportTo (new File ("./target/CIMWeb.war"), true)
            val ear = ShrinkWrap.create (classOf[EnterpriseArchive], "CIMWeb.ear")
                .addAsModules (war)
                .add (new FileAsset (new File ("../CIMConnector/target/CIMConnector-2.11-2.4.5-2.7.2.rar")), "CIMConnector.rar")
                .addManifest ()
                .addAsManifestResource (new File (WEBEAR_SRC, "application.xml"))
            println (ear.toString (true))
            ear.as (classOf[ZipExporter]).exportTo (new File ("./target/CIMWeb.ear"), true)
            //        final EnterpriseArchive ear = ShrinkWrap.createFromZipFile (EnterpriseArchive.class, new File ("../CIMEar/target/CIMApplication.ear"));
            //        println (ear.toString (true));
            //        ear.as (ZipExporter.class).exportTo (new File ("./target/CIMWeb.ear"), true);
            Some (ear)
        }
        catch
        {
            case e: Exception =>
                println (e.getMessage)
                None
        }
    }
}

@RunWith (classOf[Arquillian])
class CIMWebTest
{
    def print_context_r (context: InitialContext, name: String, depth: Int): Unit =
    {
        try
        {
            val values = context.list (name)
            while (values.hasMore)
            {
                val pair = values.next
                var i = 0
                while (i < depth)
                {
                    print ("    ")
                    i += 1
                }
                println (pair.getName + " : " + pair.getClassName)
                if ("org.apache.openejb.core.ivm.naming.IvmContext" eq pair.getClassName)
                    print_context_r (context, name + "/" + pair.getName, depth + 1)
            }
        }
        catch
        {
            case _: NamingException =>
        }
    }

    def print_context (context: InitialContext, name: String): Unit =
    {
        println (name)
        print_context_r (context, name, 1)
    }

    def propertiesFromMap (properties: Map[String, String]): Properties =
        properties.foldLeft (new Properties)
        {
            case (p, (k, v)) =>
                val _ = p.put (k,v)
                p
        }

    @Test
    def testResourceAdapter (): Unit =
    {
        val properties = Map[String, String] (
            Context.INITIAL_CONTEXT_FACTORY -> "org.apache.openejb.client.LocalInitialContextFactory",
            "openejb.deployments.classpath.include" -> ".*resource-injection.*"
        )
        val initialContext = new InitialContext (propertiesFromMap (properties))
        print_context (initialContext, "java:")
        print_context (initialContext, "openejb:")
        initialContext.lookup ("openejb:Resource/CIMConnector.rar") match
        {
            case connectionFactory: ConnectionFactory =>
                val connection = connectionFactory.getConnection
                assertNotNull (connection)
                connection.close ()
            case _ =>
                fail ("could not get a ConnectionFactory")
        }
    }
}
